### PR TITLE
Support Kotlin inline functions: try to access SDE from class files on disk

### PR DESCRIPTION
**Pull request by zuevmaxim**
In ClassFileLoad hook the `protection_domain` parameter is used to locate the class file on disk, where it is possible to extract SDE.

### CODE REVIEW

File: src/debugInfoAccess.cpp [hunk 1]
```diff
@@ -333,8 +382,111 @@ void DebugInfoAccessFromClassFiles::AnalyseLoadedClasses(jvmtiEnv* jvmti) {
...
+        index = _seen_classes.size();
+        _seen_classes[className] = index;
+    } else {
+        index = index_it->second;
+    }
+    _mutex.unlock();
+    return index;
+}
+
+std::string urlDecode(const char* src) {
```
Comment (nikita-nazarov) [inline src/debugInfoAccess.cpp:399]: This looks like a hack to the utf encoding problem. Discussed in DM.

File: src/fileAccess.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+#include "miniz.h"
+#include <string.h>
+#include <dirent.h>
+#include <fstream>
+#include <vector>
+
+class FileResource : public Resource {
+    std::vector<char> _buffer;
+  public:
+    FileResource(std::ifstream& bin_file) : Resource(), _buffer(std::istreambuf_iterator<char>(bin_file), std::istreambuf_iterator<char>()) {}
```
Comment (nikita-nazarov) [inline src/fileAccess.cpp:12]: Maybe it would be better to split this line?

File: src/fileAccess.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    Resource* get(const std::string& path) {
+        size_t size;
+        u8* buf = (u8*) mz_zip_reader_extract_file_to_heap(&_zip_archive, path.c_str(), &size, 0);
+        if (!buf) return NULL;
+        return new ZipFileResource(size, buf);
+    }
+};
+
+bool endsWith(const std::string& fullString, const std::string& ending) {
```
Comment (nikita-nazarov) [inline src/fileAccess.cpp:80]: This function can be static.

File: src/debugInfoAccess.cpp [hunk 4]
```diff
@@ -333,8 +382,111 @@ void DebugInfoAccessFromClassFiles::AnalyseLoadedClasses(jvmtiEnv* jvmti) {
...
+    _mutex.unlock();
+}
+
+void DebugInfoAccessFromClassFiles::processClassFiles() {
+    _mutex.lock();
+    if (_class_files.empty()) {
+        _mutex.unlock();
+        return;
+    }
+    std::map<std::string, std::vector<std::string> > class_files;
```
Comment (nikita-nazarov) [inline src/debugInfoAccess.cpp:461]: `std::map<std::string, std::vector<std::string> > class_files(std::move(_class_files))` ?

**Review by nikita-nazarov — CHANGES_REQUESTED** [review]

File: src/debugInfoAccess.cpp (hunk 4 above)
Comment (zuevmaxim) [reply src/debugInfoAccess.cpp:461]: Looks like `std::move` is unavailable until c++11: https://en.cppreference.com/w/cpp/utility/move

File: src/debugInfoAccess.cpp (hunk 1 above)
Comment (zuevmaxim) [reply src/debugInfoAccess.cpp]: This method is replaced with java `URLDecoder#decode` call

**Review by nikita-nazarov — APPROVED** [review]

**Pull request closed without merging**